test runner rework

tavip opened the pull request:
Rework the test runner from a Makefile based one to a python script
that generates clean test output summary similar with the Linux
kernel build output. This makes it easier to understand what tests
have failed and why when run from the command line.

The runner can also generate junit xml files that we can integrate
with the CI to generate test reports per PR.

This PR also reworks the build system by moving the autoconfiguration bits to a separate Makefile and by generating persistent host configuration: 

- Makefile.conf: stores system configuration like CFLAGS, LDFLAGS, LDLIBS, etc.
- include/autoconf.h: header file that allows build time identification if certain features are available
- tests/autoconf.sh: shell variables that allows the test runner to identify if certain features are available


<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/lkl/linux/400)
<!-- Reviewable:end -->

thehajime commented on tools/lkl/tests/net-test.c:
```diff
@@ -44,41 +77,59 @@ in_cksum(const u_short *addr, register int len, u_short csum)
 		return TEST_FAILURE;
 	}
 
 	icmp = malloc(sizeof(struct icmphdr *));
```
maybe it's okay but `sizeof(struct lkl_icmphdr *)` instead ?

tavip replied:
Yes, good catch, I'll fix it.

tavip commented on tools/lkl/Makefile:
```diff
@@ -2,192 +2,89 @@
+  OUTPUT := $(OUTPUT)/tools/lkl/
 else
-OUTPUT := $(CURDIR)/
+ OUTPUT := $(CURDIR)/
```
Good catch, I will fix it.

tavip commented on tools/lkl/tests/boot.c:
```diff
@@ -598,84 +310,31 @@ static int test_getdents64(char *str, int len)
-		if (wr >= len)
-			break;
+		wr += lkl_test_logf("%s ", de->d_name);
+		if (wr >= 70) {
```
Yes, good idea, will fix it.

tavip commented on tools/lkl/tests/boot.c:
```diff
@@ -756,34 +409,34 @@ static int test_syscall_thread(char *str, int len)
+			lkl_test_logf("write error: %s\n", lkl_strerror(ret));
 		else
-			snprintf(str, len, "write: short write: %ld", ret);
+			lkl_test_logf("short write: %ld", ret);
```
Nice catch, will fix it.

tavip commented on tools/lkl/tests/boot.c:
```diff
@@ -715,15 +372,13 @@ static int test_semaphore(char *str, int len)
 	lkl_host_ops.sem_free(sem);
 
-	snprintf(str, len, "%ld", ret);
-
```
There is not useful information to print, this tests always printed TEST_SUCCESS (and the tests status is printed anyway).

thehajime replied:
okay

tavip commented on tools/lkl/tests/hijack-test.sh:
```diff
@@ -1,591 +1,741 @@
+    echo "$addr" | grep "$(ip6_lkl)"
+    ! echo "$addr" | grep "WARN: failed to free"
+}
 
```
yes, the new command line tests is platform independent and we should not have any critical message printed to screw it up.

tavip commented on tools/lkl/tests/hijack-test.sh:
```diff
@@ -1,591 +1,741 @@
+    set -e
+
+    if [ -n "$LKL_HOST_CONFIG_ANDROID" ]; then
+        return $TEST_SKIP
```
It looks like Android does not support the standard qdiscs, here is the error I am getting:

+ LKL_HIJACK_CONFIG_FILE=/data/local/tmp/lkl/hijack-test.conf /data/local/tmp/lkl/bin/lkl-hijack.sh tc -s -d qdisc show
Android does not support qdisc 'noqueue'
Android does not support qdisc 'fq'
qdisc noqueue 0: dev lo root refcnt 2
Sent 0 bytes 0 pkts (dropped 0, overlimits 0)
qdisc fq 8001: dev eth0 root refcnt 2 [cannot parse qdisc parameters]
Sent 0 bytes 0 pkts (dropped 0, overlimits 0)

We can still probably fix this, but it needs more investigation.

thehajime replied:
okay, agree.

thehajime commented on tools/lkl/tests/test.sh:
```diff
@@ -0,0 +1,174 @@
+    echo "$@" | $WRAPPER sh -x
+}
+
+adb_push()
```
maybe we can remove/delete tools/lkl/tests/android-test.sh ?

thehajime replied:
`make test` could be replaced as well

tavip replied:
Yes, these are removed in the last patch of the series.

thehajime replied:
Thanks,

in tools/lkl/Makefile there is still the target `test valgrind gdb` which calls `make -C tests`.

https://github.com/lkl/linux/pull/400/files#diff-bb2bc012ca931e05448c2b62f99c55abR107

So removing tools/lkl/tests/Makefile lets those targets orphaned while removing android-test.sh is okay.

I think `make gdb/tests/valgrind` is still useful so how about keeping `tools/lkl/tests/Makefile` ?

tavip replied:
For gdb and valgrind, I think we should add an env var to select the proper wrapper in test.sh. That way we can run gdb/valgrind for all commands not only boot. 

As for the tests target, I guess we can add a target into the top Makefile to run tests/run.py.

I'll do these changes in the new version I'll push soon.

thehajime replied:
thanks too.  I agree with what you're going to update.

thehajime commented on tools/lkl/tests/hijack-test.sh:
```diff
@@ -1,591 +1,741 @@
+    tap_cleanup 1
+}
+
+test_tap_ping_host()
```
I don't know how it works on android.
IIRC android doesn't have tap support but the following test results says `TEST       ok       test_tap_ping_host `.  

https://circleci.com/gh/tavip/linux/134?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

tavip replied:
It does work, it just needs the right permissions, which in this case are: user=vpn,group=vpn and additional group=net_admin. See the tap_prepare and tap_setup in net-setup.sh

thehajime replied:
that's new to me (cc: @phhusson @linhua55 and #379). Great and thanks !

tavip commented:
The checkpatch error is about using __attributte__(printf) instead of the __printf define in the kernel headers. There is a definition in tools/include but including those headers pulls in other dependincies that are failing on some platfors, so I think it is sensible for now to just go with __attribute__.

thehajime commented on tools/lkl/tests/net.sh:
```diff
@@ -1,102 +1,175 @@
+    "loopback")
+        ;;
+    "pipe")
+        if [ -z $(lkl_test_cmd which mkfifo) ]; then
```
do you see why x86_64 test skips pipe test ?

https://circleci.com/gh/tavip/linux/151?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link

tavip replied:
There was an issue in the test, I've fixed it and will push a new version soon.

thehajime replied:
thanks, will look at your new version.

thehajime commented:
> The checkpatch error is about using attributte(printf) instead of the __printf define in the kernel headers. There is a definition in tools/include but including those headers pulls in other dependincies that are failing on some platfors, so I think it is sensible for now to just go with attribute.

Okay, agree.

tavip commented:
@thehajime Fixed the pipe test. Also added support for gdb and valgrind by either using --gdb or --valgrind options with the test runner or by setting VALGRIND or GDB env vars to non-null values.

Also added nightly build that runs valgrind on x86_64.

thehajime commented:
Thanks, valgrind nightly build needs a bit more work but I will create another PR for that.
